### PR TITLE
Define and implement new predicates.hpp APIs

### PR DESCRIPTION
**Pull request by cwharris**

### CODE REVIEW

**Comment by cwharris** [comment]
Seems like this API should be moved to `sorting.hpp` ?

File: cpp/src/predicates/is_sorted.cu [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+bool is_sorted(cudf::table_view const& in,
+               std::vector<order> const& column_order,
+               std::vector<null_order> const& null_precedence)
+{
+    if (in.num_columns() == 0 || in.num_rows() == 0) {
+        return true;
+    }
+
+    if (not column_order.empty()) {
```
Comment (cwharris) [inline cpp/src/predicates/is_sorted.cu]: What if we have more or less than one column in the table and we have only one item in `column_order` and/or `null_precedence`? Should we use the same `CUDF_EXPECTS`, or assume those orders for all columns?

Comment (jrhemstad) [reply cpp/src/predicates/is_sorted.cu]: I'm not following. The accepted states for `column_order` and `null_precedence`:
```
column_order.empty() or column_order.size() == in.num_columns()

null_precedence.empty() or null_precedence.size() == in.num_columns()
```

Anything else is an error.

Comment (cwharris) [reply cpp/src/predicates/is_sorted.cu]: Is it worth adding any overload of `is_sorted` which takes a single `order` / `null_order`, and generates an appropriately sized vector, initialized to the given `order` / `null_order`?

Comment (jrhemstad) [reply cpp/src/predicates/is_sorted.cu]: It's trivial to do this at the call site:

```
is_sorted(t); // assumes all columns ascending
is_sorted(t, std::vector<order>(ASCENDING, t.num_columns())); // this is the same thing as above
is_sorted(t, std::vector<order>(DESCENDING, t.num_columns()); // all columns descending
```

File: cpp/include/cudf/predicates.hpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/*
```
Comment (jrhemstad) [inline cpp/include/cudf/predicates.hpp]: I agree that this could probably move to `sort.hpp` and it would still make sense.

File: cpp/tests/orderby/is_sorted/is_sorted_tests.cu [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    std::vector<cudf::order> order{ };
+    std::vector<cudf::null_order> null_precedence{ cudf::null_order::BEFORE,
+                                                   cudf::null_order::BEFORE };
+
+    EXPECT_THROW(cudf::experimental::is_sorted(in, order, null_precedence),
+                 cudf::logic_error);
+}
+
+template <typename T>
+struct IsSortedNvstringCategory : public cudf::test::BaseFixture {};
```
Comment (jrhemstad) [inline cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: NVStringCategory doesn't exist anymore. However, could you add some tests that include string columns? You can use `strings_column_wrapper` to help with creating/initializing columns of strings.

File: cpp/tests/orderby/is_sorted/is_sorted_tests.cu [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+#include <vector>
+
+
+template<typename T>
+using fixed_width_column_wrapper = cudf::test::fixed_width_column_wrapper<T>;
+
+template <typename T>
+struct IsSortedNumeric : public cudf::test::BaseFixture {};
+
+using test_types_numeric = ::testing::Types<int8_t, int16_t, int32_t, int64_t,
```
Comment (jrhemstad) [inline cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: This should use one of the type lists like `cudf::test::FixedWidthTypes`.

File: cpp/include/cudf/predicates.hpp (hunk 2 above)
Comment (cwharris) [reply cpp/include/cudf/predicates.hpp]: @harrism Any objections to moving `cudf::is_sorted` from `predicates.hpp` to `sorting.hpp` (specifically: as part of this PR)?

Comment (harrism) [reply cpp/include/cudf/predicates.hpp]: Nope. I assume there are no other predicates?

**Comment by harrism** [comment]
rerun tests

File: cpp/tests/orderby/is_sorted/is_sorted_tests.cu (hunk 4 above)
Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: @jrhemstad ... what if i wrote tests that work for both NumericTypes and std::string? would I have to duplicate those test definitions?

Comment (harrism) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: `cudf::test::AllTypes`?

Comment (jrhemstad) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: `AllTypes` is tricky because you can't just uniformly use `fixed_width_column_wrapper<TypeParam>`, if `TypeParam == string_view` you need to use `string_column_wrapper`.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: I have tests which will work for both `fixed_width_column_wrapper` and `string_column_wrapper`, the only difference being:

```
TYPED_TEST_CASE(IsSortedNumeric, cudf::test::NumericTypes);
```
```
TYPED_TEST_CASE(IsSortedNumeric, ::testing::Types<std::string>);
```

I'm accomplishing this by injecting test data via explicit specialization. Syntactically, the tests for all data types are the same - the only difference being the data used in those tests.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: This approach also allows me to avoid special exceptions for specific types, like the one seen in the `sorted_order` tests when using `::experimental::bool8`.

https://github.com/rapidsai/cudf/blob/branch-0.11/cpp/tests/sort/sort_test.cpp#L52-L68

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: I did a quick experiment to see what would happen when I removed fixed_width_column's static_assert for fixed-width types (which is what was preventing me from compiling). The code compiles, but some of my tests use empty vectors, and apparently our string column factory doesn't support zero length string columns... those tests failed, and I stopped digging at that point.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: I pushed the tests so you have can have a better idea what I'm talking about.

File: cpp/tests/orderby/is_sorted/is_sorted_tests.cu [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+                                    std::string>;
+
+// compiles and passes tests
+TYPED_TEST_CASE(IsSortedNumeric, cudf::test::NumericTypes);
+
+// compiles and passes tests
+// TYPED_TEST_CASE(IsSortedStrings, ::testing::Types<std::string>);
+
+// // compiles and passes tests if static_asserts for fixed-width types are removed
+// TYPED_TEST_CASE(IsSortedNumeric, test_types);
```
Comment (jrhemstad) [inline cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: Instead of specializing for `std::string`, use `cudf::string_view` and use the `cudf::test::AllTypes`. Then it should work.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: `cudf::test::AllTypes` doesn't include `string_view`, but it does include `TimestampTypes`, which these tests do not support. I've updated to this which works flawlessly:
```
using test_types = Concat<NumericTypes, Types<cudf::string_view>>;
TYPED_TEST_CASE(IsSortedTest, test_types);
```

Comment (jrhemstad) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: > 1) add string_view to AllTypes

I think this makes sense.

> 2) make these tests work with timestamps 

Why don't they support timestamps? As far as sorting is concerned, I'd think they'd be no different from a numeric type.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: The legacy tests did not support time stamps, so I didn't add them (and they compiled but did not pass - now that I'm thinking about it, it might be due negative integers). Also, I didn't see any examples on how to use timestamps so far. At least not in tests. I'll take another shot at it.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: It may be worth adding `LexicographicalTypes`.

Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: Adding `string_view` to `AllTypes` breaks other tests (expected). I've added `LexicographicalTypes`, and I'm using that in the tests instead. `AllTypes` should have `string_view`, but other tests need updating for that to happen. Advice?

File: cpp/tests/utilities/type_lists.hpp [hunk 6]
```diff
@@ -101,11 +113,23 @@ using TimestampTypes = cudf::test::Types<timestamp_D, timestamp_s, timestamp_ms,
...
+ * @brief Provides a list of all fixed-width element types for use in GTest
+ * typed tests.
+ * 
+ * Example:
+ * ```
+ * // Invokes all typed fixture tests for all fixed-width types in libcudf
+ * TYPED_TEST_CASE(MyTypedFixture, cudf::test::FixedWidthTypes);
+ * ```
+ *---------------------------------------------------------------------------**/
+using LexicographicalTypes = Concat<NumericTypes, TimestampTypes, StringTypes>;
```
Comment (jrhemstad) [inline cpp/tests/utilities/type_lists.hpp]: Instead of adding a new type list, why not add `StringTypes` to `AllTypes`?

Comment (cwharris) [reply cpp/tests/utilities/type_lists.hpp]: Practically, it breaks other tests, and I feel fixing those tests falls outside the scope of this PR. Conceptually, it's conceivable not all types will be lexicographical, and therefore having a type list specific to lexicographical types prevents us from having to changes these tests in the future if a non-sortable type is added to `AllTypes` (which is similar to the problem I'm seeing when adding `StringTypes` to `AllTypes`.

TLDR; `StringTypes` should probably be added to `AllTypes`, but it would require making sure some unrelated tests are updated.

Comment (jrhemstad) [reply cpp/tests/utilities/type_lists.hpp]: Makes sense. I'd just suggest renaming `LexicographicalTypes` to `ComparableTypes`.

File: cpp/tests/orderby/is_sorted/is_sorted_tests.cu (hunk 5 above)
Comment (cwharris) [reply cpp/tests/orderby/is_sorted/is_sorted_tests.cu]: https://github.com/rapidsai/cudf/pull/3299#discussion_r345354714

File: cpp/tests/utilities/type_lists.hpp (hunk 6 above)
Comment (cwharris) [reply cpp/tests/utilities/type_lists.hpp]: Done.

File: cpp/include/cudf/predicates.hpp (hunk 2 above)
Comment (cwharris) [reply cpp/include/cudf/predicates.hpp]: Correct.
